// @docusaurus/plugin-pwa/lib/options.js
"use strict";
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateOptions = exports.Schema = void 0;
const tslib_1 = require("tslib");
const path_1 = (0, tslib_1.__importDefault)(require("path"));
const utils_validation_1 = require("@docusaurus/utils-validation");
const DEFAULT_OPTIONS = {
    debug: false,
    offlineModeActivationStrategies: [
        'appInstalled',
        'queryString',
        'standalone',
    ],
    injectManifestConfig: {},
    pwaHead: [],
    swCustom: undefined,
    swRegister: path_1.default.join(__dirname, 'registerSw.js'),
    reloadPopup: '@theme/PwaReloadPopup',
};
exports.Schema = utils_validation_1.Joi.object({
    debug: utils_validation_1.Joi.bool().default(DEFAULT_OPTIONS.debug),
    offlineModeActivationStrategies: utils_validation_1.Joi.array()
        .items(utils_validation_1.Joi.string()
        .valid('appInstalled', 'queryString', 'standalone', 'mobile', 'saveData', 'always')
        .required())
        .default(DEFAULT_OPTIONS.offlineModeActivationStrategies),
    injectManifestConfig: utils_validation_1.Joi.object().default(DEFAULT_OPTIONS.injectManifestConfig),
    pwaHead: utils_validation_1.Joi.array()
        .items(utils_validation_1.Joi.object({ tagName: utils_validation_1.Joi.string().required() }).unknown().required())
        .default(DEFAULT_OPTIONS.pwaHead),
    swCustom: utils_validation_1.Joi.string(),
    swRegister: utils_validation_1.Joi.alternatives()
        .try(utils_validation_1.Joi.string(), utils_validation_1.Joi.bool().valid(false))
        .default(DEFAULT_OPTIONS.swRegister),
    reloadPopup: utils_validation_1.Joi.alternatives()
        .try(utils_validation_1.Joi.string(), utils_validation_1.Joi.bool().valid(false))
        .default(DEFAULT_OPTIONS.reloadPopup),
});
function validateOptions({ validate, options, }) {
    return validate(exports.Schema, options);
}
exports.validateOptions = validateOptions;
